### PR TITLE
Do not ignore missing imports in mypy run by default

### PR DESCRIPTION
**Pull request by zundertj**
Under the global config `ignore_missing_imports`, if mypy does not detect a `py.typed` file, it will use stubs, but only if available, for all our dependencies. Otherwise it will assume `Any` as types for all dependencies, meaning no type hinting.

In this PR:
* remove the global `ignore_missing_imports`
* add dependency pandas-stubs
* for dependencies that gave errors and could not easily find stubs, added to ignore list (see `pyproject.toml`. We can iterate on this in the future
* make changes to code base to deal with type errors that were not noticed before due to this global config

### CODE REVIEW

**Comment by zundertj** [comment]
Ok, CI run failing because for Py3.6 it install a numpy version < 1.20, which is the version that introduced typing on the codebase & `py.typed` file. Numpy 1.20 does not support Python 3.6: https://numpy.org/devdocs/release/1.20.0-notes.html.

**Comment by ritchie46** [comment]
Ok, so as discussed in #2188, I think we can update the github action to now use python 3.7 as minimal python version and run this again.

File: py-polars/pyproject.toml [hunk 1]
```diff
@@ -22,9 +22,12 @@ connectorx = ["connectorx"]
 profile = "black"
 
 [tool.mypy]
-ignore_missing_imports = true
 disallow_untyped_defs = true
 files = ["polars", "tests"]
 
+[[tool.mypy.overrides]]
```
Comment (ritchie46) [inline py-polars/pyproject.toml:28]: So the idea is that when these libs get stubs we can include them one by one?

Comment (zundertj) [reply py-polars/pyproject.toml:28]: Yes, either stubs (such as Pandas) or add a `py.typed` file to denote the type annotations in the code can be used (like numpy, and polars itself)

Comment (ritchie46) [reply py-polars/pyproject.toml:28]: Oh. so libraries can have different tactics? We added a `py.typed`, but you can also create a separate `stubs` package?

**Pull request merged**